### ISSUE TITLE
Multi Unit Item

### ISSUE THREAD

**Issue by andreievg** [msupply-foundation/open-msupply#2078]
From KDD: https://github.com/openmsupply/open-msupply/pull/1918

# As Pack Lens - Option 3

Allow item base unit to be viewed as one of centrally configured unit variants, quantities will be displayed based on the selected unit variant and display pack size and unit information based on those configured variants. Below diagram describes it well, [also in google docs](https://docs.google.com/spreadsheets/d/1mWZqmfQRfHlwF5i2OADaBaSWalK3FbIJpauM275xYuA/edit#gid=937944726):
 
![pack_lens (1)](https://github.com/openmsupply/open-msupply/assets/26194949/bb75dc76-786b-41b7-a7fd-f29535ff63a0)

# Tasks

- [x] https://github.com/openmsupply/open-msupply/issues/2147
     * Mock API - **4hours**
     * Provider/context and hook - **8hours**
     * Conversion from pack size to unit variant in stock view - **2hours**
- [x]  https://github.com/openmsupply/open-msupply/issues/2156 (~~including storing to app data~~ and memorisation within the hook), add to catalog view - **4hours**
- [x] Store user selection in app data
- [x] https://github.com/openmsupply/open-msupply/issues/2208 - **8hours**
- [x] https://github.com/openmsupply/open-msupply/pull/2149#discussion_r1309389058
- [x] https://github.com/openmsupply/open-msupply/issues/2176
- [x] Add enterable field based on selected unit pack (for internal orders and outbound shipment inputs) 
- [x] https://github.com/openmsupply/open-msupply/issues/2313

Then use above in

- [x] Stocktake - **1hours**
- [x] Inbound Shipment - **1hours**
- [x] Requisition - **1hours**
- [x] Internal Order - **4hours**
- [x] Above 4 in this issue: https://github.com/openmsupply/open-msupply/issues/2214 
- [x] Refine outbound shipment issue modal UI (meeting with team to work it out, ask Anisha for help with this) - ?
- [x] Outbound Shipment - **8hours** (to do basic one, with columns the way they are in the diagram)
- [x] https://github.com/openmsupply/open-msupply/issues/2309 - **8hours**

Back End

- [x] https://github.com/openmsupply/open-msupply/issues/2310

Central

- [ ] Sync -> depending on how we want to do central (mSupply or parallel omSupply, estimate for omSupply central) https://github.com/openmsupply/open-msupply/issues/2062 - **24hours** 
- [ ] Sync progress (make issue and discussion issue)

Extra

- [x] https://github.com/openmsupply/open-msupply/issues/2314
- [ ] What about decimals and example of how it works for say mills

Documentation

- [ ] Internal (with specs, also may need to simplify naming and overall comments: https://github.com/openmsupply/open-msupply/pull/2149#discussion_r1308208153) https://github.com/openmsupply/open-msupply/issues/2312 - **2hours**
- [ ] https://github.com/openmsupply/open-msupply/issues/2311- **2hours**

# Break down

From end consumer perspective, we have three types of fields that would need to be adjusted by the unit/pack lens. We can use these methods from a hook to address consumer requirements:

### Pack/Unit column

![Screenshot 2023-08-08 at 3 22 58 PM](https://github.com/openmsupply/open-msupply/assets/26194949/9a2d91cc-fa83-401d-b5c8-b4ceceea512d)

```ts
const { asPackUnit } = useUnitVariant("itemIdHere");
// .. 

// For this item there is a unit variant of packSize 10 with name = 'blist of 10 tab'
const packSize = 10;
const packUnit = asPackUnit(packSize); // i.e. blist of 10 tab

// For this item there is no variant matching packSize 10. Unit name is tablet
const packSize = 10;
const packUnit = asPackUnit(packSize); // i.e. tablet x 10
```

### Quantity column

![Screenshot 2023-08-08 at 3 59 25 PM](https://github.com/openmsupply/open-msupply/assets/26194949/490a761a-ac13-446f-a4ad-fc5f7120791d)

For all quantity fields above

```ts
const { numberOfPacksFromQuantity } = useUnitVariant("itemIdHere");
// ..
const totalQuantity = 1000;
const packUnit = numberOfPacksFromQuantity(totalQuantity);
```

### Variant Selector

![Screenshot 2023-08-08 at 4 02 20 PM](https://github.com/openmsupply/open-msupply/assets/26194949/c47f098c-273f-4626-99c0-f8265ebf07d7)

To build Pack/Unit drop down we would need

```tsx
const { options, setDefaultOption, defaultOption } = useUnitVariant("itemIdHere");
// ..
<>
  <dropdown options={options} default={defaultOption} onChange={(newOption) => setDefaultOption(newOption)}/>
</>
```

**NOTE:** setDefaultOption is most likely a bad name in this context, user is choosing their preferred option no the default option

## The hook and context

For the easy of use, we would define a top level provider/context that fetches multi unit variants and default variant (based on stock lines) from API and combines default with app data (if user select variant other then calculated default variant from back end, it is saved in app data as default), and then a hook is exposed

```ts
// I think having the "itemId" as parameter will allow for easier memorisation 
const { asPack, numberOfPacksFromQuantity, options, setDefaultOption, defaultOption } = useItemUnitVariants("itemId")
```

A data format that would be easy for hook to use would be like this:

```ts
const items = [
  {
    itemId: 'uuidhere',
    unit: {},
    // Back end determines (base on stock lines in stock ? Or based on number of stock lines)
    mostUsed: 'idHere',
    // Use can overwrite and store in app data
    userSet: 'idHere',
    variants: [
      {
        shortName: 'blist of 10',
        longName: 'blister of 10 tablets',
        id: 'idHere',
        packSize: 10,
      },
    ],
  },
];

```

By memorisation I was mean that we reduce number of re-renders when one unit variant default/selection is changed (making sure we only update fields for the one that was changed), this should be simple enough.

## API Result

I think API should reply with data that is very close to the consumer data:

```ts
const items = [
  {
    itemId: 'uuidhere',
    unit: {},
    // Back end determines (base on stock lines in stock ? Or based on number of stock lines)
    mostUsed: 'idHere',
    //  userSet: 'idHere', -> this is an extension that is populated by the provider/hook/context
    variants: [
      {
        shortName: 'blist of 10',
        longName: 'blister of 10 tablets',
        id: 'idHere',
        packSize: 10,
      },
    ],
  },
];
```

## App data

Storage of user picked unit variant should be pretty simple, just key value, key = itemId and value = unitVariantId

## Schema and configurations

I was hoping we could do configurations in parallel [omSupply Central server](https://github.com/openmsupply/open-msupply/issues/2062), I estimate the base functionality would be 3 or so days of work.

We can of course do the configuration as global pref in JSON form on mSupply server for now.

Btw schema would just need to be able to store this: 

![Screenshot 2023-08-08 at 4 54 36 PM](https://github.com/openmsupply/open-msupply/assets/26194949/1fadeaa9-4bce-446e-8aa7-f272b2cc10f6)

## MW job code 
OMS:NEST

**Comment by andreievg** [comment]
**Question 1:**

How do we figure out what is the default (if not chosen by user):

1 - Number of stock lines (all together, including 0 stock lines) ?
2 - Most quantity of stock lines (non zero ones ofcourse)

2 and then 1 ?

**Comment by andreievg** [comment]
**Question 2:**

For internal order, we need to adjust all of the graphs, should this be done:

1 - On front end
2 - Provide unit variant to API

**Comment by andreievg** [comment]
**Question 3:**

In the KDD i went a little bit over the board with outbound shipment modifications:

![Screenshot 2023-08-08 at 5 15 49 PM](https://github.com/openmsupply/open-msupply/assets/26194949/e36cd1c9-7c38-4dd1-b75d-2054cb9a951c)

I think for now maybe we shouldn't relocate placeholder etc, but I think having the grouping of pack and unit could help (and unit variant selector for placeholder).

Was wondering what your thoughts are on above view in general vs what we have now

**Comment by andreievg** [comment]
**Question 4:**

Do we go with current mSupply and json config or omSupply central server in parallel and do UI configurations there, as per [this issue 
](https://github.com/openmsupply/open-msupply/issues/2062), I think it's about 3 days of work. I need to summarise my finding from RnD day a bit further, but yeah central data config should be pretty simple.

**Comment by mark-prins** [comment]
thanks @andreievg! that's a very comprehensive write up. an epic epic.
nicely covered - can't speak for anyone else of course, but I think that you've captured what's required / wanted. I do like the kdd option 3 (tho will go back and re-read the latest version to double check)

had some questions after reading through:

for this example - wouldn't the result be `tablet x 10`? 
```
const { asPackUnit } = useUnitVariant("itemIdHere");
// ..
const packSize = 10;
const packUnit = asPackUnit(packSize); // i.e. blist of 12 tab
```

If so, the comment is potentially confusing things.
If not - what happens when you have a `packSize` which doesn't match any of the available variants?


I'm concerned about being able to specify a placeholder quantity in a random pack size, which would result in non-integer numbers of placeholder quantities. Are we to allow part-packs of placeholders?

with `const { options, setDefaultOption, defaultOption } = useUnitVariant("itemIdHere");` I think `setDefaultOption` is a bad idea: the default option is just that, the 'default' to display before a user touches the control. typically there is a 'default' and a 'selected'. The user isn't changing the default option, they are changing the selected option.

for the hook, are you suggesting that we fetch _all_ items from the server and keep those in the context? tbh I can't see how else we could do it, as we need to have all items available and fetching each item when requested is going to be too slow I think.

..and some answers to your questions:

Question 1: yes, I like 2 then 1
Q2: front end should be ok - it's not a lot of data. I'd be concerned at the speed of a round trip to the server every time the pack selection changes
Q3: I am a little biased, having spent a long time reworking that modal already. I like the placeholder at the bottom, don't have a problem with 'Any' and was confused with the 'unit' section - since the price shown is per unit, but the quantity is a total, but expressed in units. Also not sure if there's enough space on screen for everything.

basically - yes, we can improve. Would prefer to leave that with Anisha and some PM type people to work through the process this time.

Q4: I rather like the idea to have a new central server implementation 😉 

🙏

**Comment by andreievg** [comment]
@mark-prins thanks for your feedback.

> for this example - wouldn't the result be tablet x 10?
> If so, the comment is potentially confusing things.
> If not - what happens when you have a packSize which doesn't match any of the available variants?

I could have done better in clarifying this:

If there is a matching variant (matched by pack size), then we show the `short name` for that variant, if there is no matching variants, then it's `{unit} x {pack size}` (or just `{unit}` if pack size is 1 and no variants with packs size 1 are defined), i refined that example, btw this is kind of what it would look like.
![Screenshot 2023-08-09 at 11 51 27 AM](https://github.com/openmsupply/open-msupply/assets/26194949/c6077ff7-4a81-4cf5-aa84-0c1f6ad3f14c)

Does that make sense ?

> I'm concerned about being able to specify a placeholder quantity in a random pack size, which would result in non-integer numbers of placeholder quantities. Are we to allow part-packs of placeholders?

The `lens` for placeholder line is just for display purpose really. Behind the scene placeholder would still be saved with packSize 1 for now (so should always be whole number). Although I agree that that view might need a bit more thought, I've made a sub task "Refine outbound shipment issue modal UI (meeting with team to work it out)"

> with const { options, setDefaultOption, defaultOption } = useUnitVariant("itemIdHere"); I think setDefaultOption is a bad idea: the default option is just that, the 'default' to display before a user touches the control. typically there is a 'default' and a 'selected'. The user isn't changing the default option, they are changing the selected option.

Nice, agree, added a comment in main description that it's not default that user is choosing but their preferred, implemented should adjust accordingly.

This brings another question, will be added below. 

> for the hook, are you suggesting that we fetch all items from the server and keep those in the context? tbh I can't see how else we could do it, as we need to have all items available and fetching each item when requested is going to be too slow I think.

It would be like a global configuration, that we don't actually need to 'wait' for,  the query should be pretty quick and the multi unit variant configuration is not compulsory (don't have to use it and if there is no config it should be pretty fast.

> Q3: I am a little biased, having spent a long time reworking that modal already. I like the placeholder at the bottom, don't have a problem with 'Any' and was confused with the 'unit' section - since the price shown is per unit, but the quantity is a total, but expressed in units. Also not sure if there's enough space on screen for everything.

Oki, i think another discussion is needed there, as per the created sub task, one thing i've noticed with omSupply UI (and of course even more so with mSupply), is that columns appear in different order in different places, i think the entry box for say pack qty to issue should be quite close to the pack size, and typically a column that's auto updated or column that is total (in this case it's both, total qty) should appear after the data entry column (pack qty), that was one of the reasons for re-arrangment 

> basically - yes, we can improve. Would prefer to leave that with Anisha and some PM type people to work through the process this time.

Great idea to get Anisha on board, I've changed the sub task slightly to mention this.

Thanks Mark!

**Comment by andreievg** [comment]
**Question 5**

Since the preferred option chose by user will trump the default option, do we need to have a way to clear the preferred option chosen by user ? I think it should be fine for user to choose the option and for it to persist for ever (or until app data is cleared, etc)

**Comment by mark-prins** [comment]
> Does that make sense ?

Yep - the explanation next the example is clearer now, and it helps having the extra example 🙏  ( though in the revised example, a pack size `10` matches to `blist of 12 tab` which would probably be 12 tablets and not 10.. I understand what you mean though! )

> ( that comment about the outbound shipment 'add item' modal and the columns and such.. )

totally! I had a mental note to review all the column names, you're right about the inconsistent order of columns too. We need to have a bit of a tidy

**Comment by andreievg** [comment]
> 10 matches to blist of 12 tab

🤦‍♂️

**Comment by andreievg** [comment]
If unit and pack size columns are combined, we may loose info about unit of item (if no variants exists so nothing is given by the server), this can be dealt with by making sure that server return unit/item or by updating ItemVariantPackUnit to also have unitName extractor and passing that as back up to asPackUnit

**Comment by andreievg** [comment]
What about a use case where there is a pack size of 100. But there are actually two variants, one is box of 10 blisters of 10 and box of 4 blisters of 25. We can't really capture this in any way, unless new fields are added. I think this use case needs to be communicated to the users, just noting it here to be captured in the docs.

**Comment by roxy-dao** [comment]
Issues have been made for remaining points.

**Comment by marthakeezy** [comment]
This has been tested as a part of V2.0.0 Release, so not sure why it's still in our QA board... Moving this and related issues to Done ✅
